port/osx/omrosdump.c

```diff
@@ -0,0 +1,473 @@
+			if (KERN_SUCCESS == kr) {
+				mach_vm_deallocate(task_port, data_read, data_size);
+				*segment_count += 1;
+			}
+		}
+		address += size;
+	}
+	if (KERN_INVALID_ADDRESS == kr) { /* from reaching end of memory in vm_region_recurse_64 */
+		kr = KERN_SUCCESS;
+	}
+
+done:
+	if (KERN_SUCCESS == kr) {
+		tmp_segments = calloc(*segment_count, sizeof(struct segment_command_64));
+		if (NULL == tmp_segments) {
+				kr = KERN_NO_SPACE;
+		} else {
+			memcpy(tmp_segments, segments, *segment_count * sizeof(struct segment_command_64));
+			*segment_commands = tmp_segments;
+		}
+		free(segments);
+	} else {
+		if (NULL != segments) {
+			free(segments);
+		}
+		if (NULL != tmp_segments) {
+			free(tmp_segments);
+		}
+	}
+	return kr;
+}
+
+
+/**
+ * Create a dump file of the OS state.
+ *
+ * @param[in] portLibrary The port library.
+ * @param[in] filename Buffer for filename optionally containing the filename where dump is to be output.
+ * @param[out] filename filename used for dump file or error message.
+ * @param[in] dumpType Type of dump to perform. Unused on OSX
+ * @param[in] userData Implementation specific data. Unused on OSX
+ *
+ * @return 0 on success, non-zero otherwise.
+ *
+ * @note filename buffer can not be NULL.
+ * @note user allocates and frees filename buffer.
+ * @note filename buffer length is platform dependent, assumed to be EsMaxPath/MAX_PATH
+ *
+ * @note if filename buffer is empty, a filename will be generated.
+ * @note if J9UNIQUE_DUMPS is set, filename will be unique.
+ */
+uintptr_t
+omrdump_create(struct OMRPortLibrary *portLibrary, char *filename, char *dumpType, void *userData)
+{
+	pid_t parent_pid = getpid();
+	pid_t child_pid = 0;
+	char *lastSep = NULL;
+	kern_return_t kr = KERN_SUCCESS;
+	mach_port_t pass_port = MACH_PORT_NULL;
+	mach_port_t special_port = MACH_PORT_NULL;
+
+	/* set core name, defaults to "core.%u" if none given */
+	if (NULL == filename || ('\0' == filename[0])) {
+		snprintf(corefile_name, PATH_MAX, "core.%u", parent_pid);
+	} else {
+		lastSep = strrchr(filename, DIR_SEPARATOR);
+		if (NULL != lastSep) {
+			strncpy(corefile_name, lastSep + 1, PATH_MAX);
+		} else {
+			strncpy(corefile_name, filename, PATH_MAX);
+		}
+	}
+
+	/* pass parent task port to child through special port inheritance */
+	kr = task_get_bootstrap_port(mach_task_self(), &special_port);
+	if (KERN_SUCCESS != kr) {
+		mach_error("failed get special port:\n", kr);
+		return kr;
+	}
+	pass_port = mach_task_self();
+	kr = task_set_bootstrap_port(mach_task_self(), pass_port);
+	if (KERN_SUCCESS != kr) {
+		mach_error("failed set special port:\n", kr);
+		return kr;
+	}
+
+	child_pid = fork();
+	if (0 == child_pid) { /* in child process */
+		kr = task_get_bootstrap_port(mach_task_self(), &pass_port);
+		if (KERN_SUCCESS != kr) {
+			mach_error("failed get special port:\n", kr);
+			return kr;
+		}
+		kr = coredump_to_file(pass_port, parent_pid);
+		raise(SIGKILL);
```

Comment:
What is this for? Shouldn't the child just `exit(0)`?

Reply:
It's so we don't run any of the normal exit procedures in the child process.